selfdrive/locationd/torqued.py

```diff
@@ -0,0 +1,267 @@
+      try:
+        cache_ltp = log.Event.from_bytes(torque_cache).liveTorqueParameters
+        cache_CP = car.CarParams.from_bytes(params_cache)
+        if self.get_restore_key(cache_CP, cache_ltp.version) == self.get_restore_key(CP, VERSION):
+          initial_params = {
+            'latAccelFactor': cache_ltp.latAccelFactorFiltered,
+            'latAccelOffset': cache_ltp.latAccelOffsetFiltered,
+            'frictionCoefficient': cache_ltp.frictionCoefficientFiltered,
+            'points': cache_ltp.points
+          }
+          self.decay = cache_ltp.decay
+          self.filtered_points.load_points(initial_params['points'])
+          cloudlog.info("restored torque params from cache")
+      except Exception:
+        cloudlog.exception("failed to restore cached torque params")
+        params.remove("LiveTorqueCarParams")
+        params.remove("LiveTorqueParameters")
+
+    self.filtered_params = {}
+    for param in initial_params:
+      self.filtered_params[param] = FirstOrderFilter(initial_params[param], self.decay, DT_MDL)
+
+  def get_restore_key(self, CP, version):
+    a, b = None, None
+    if CP.lateralTuning.which() == 'torque':
+      a = CP.lateralTuning.torque.friction
+      b = CP.lateralTuning.torque.latAccelFactor
+    return (CP.carFingerprint, CP.lateralTuning.which(), a, b, version)
+
+  def reset(self):
+    self.resets += 1.0
+    self.invalid_values_count = 0
+    self.raw_points = defaultdict(lambda: deque(maxlen=self.hist_len))
+    self.filtered_points = PointBuckets(x_bounds=STEER_BUCKET_BOUNDS, min_points=MIN_BUCKET_POINTS)
+
+  def estimate_params(self):
+    points = self.filtered_points.get_points(FIT_POINTS_TOTAL)
+    # total least square solution as both x and y are noisy observations
+    # this is empirically the slope of the hysteresis parallelogram as opposed to the line through the diagonals
+    points = np.insert(points, 1, 1.0, axis=1)
+    _, _, v = np.linalg.svd(points, full_matrices=False)
+    slope, offset = -v.T[0:2, 2] / v.T[2, 2]
+    _, spread = np.einsum("ik,kj -> ji", np.column_stack((points[:, 0], points[:, 2] - offset)), slope2rot(slope))
+    friction_coeff = np.std(spread) * FRICTION_FACTOR
+    return slope, offset, friction_coeff
+
+  def update_params(self, params):
+    self.decay = min(self.decay + DT_MDL, MAX_FILTER_DECAY)
+    for param, value in params.items():
+      self.filtered_params[param].update(value)
+      self.filtered_params[param].update_alpha(self.decay)
+
+  def is_sane(self, latAccelFactor, latAccelOffset, friction):
+    if any([val is None or np.isnan(val) for val in [latAccelFactor, latAccelOffset, friction]]):
+      return False
+    return (((1.0 + FACTOR_SANITY) * self.offline_latAccelFactor) >= latAccelFactor >= ((1.0 - FACTOR_SANITY) * self.offline_latAccelFactor)) & \
+      (((1.0 + FRICTION_SANITY) * self.offline_friction) >= friction >= ((1.0 - FRICTION_SANITY) * self.offline_friction))
```

Comment:
can you split this up into multiple lines and use logical operators instead of bitwise